### PR TITLE
Implement DataFrame pct_change

### PR DESCRIPTION
**Pull request by skirui-source**
Fixes: https://github.com/rapidsai/cudf/issues/9603

Next steps:

- [ ] Fix https://github.com/rapidsai/cudf/issues/10314 
- [ ] Move `diff `and `pct_change` to `indexed_frame`  so that both are accessible to `Series` as well

### CODE REVIEW

File: python/cudf/cudf/core/dataframe.py [hunk 1]
```diff
@@ -6329,6 +6329,45 @@ def explode(self, column, ignore_index=False):
...
+        elif fill_method not in {"ffill", "pad", "bfill", "backfill"}:
+            raise ValueError(
+                "fill_method must be one of 'ffill', 'pad', "
+                "'bfill', or 'backfill'."
+            )
+
+        data = self.fillna(method=fill_method, limit=limit)
+        data_diff = data.diff(periods=periods)  # need to implem. diff method
+        change = data_diff / data.shift(periods=periods, freq=freq)
+        return change
```
Comment (karthikeyann) [inline python/cudf/cudf/core/dataframe.py:6369]: skip these intermediates.
```suggestion
        return data.diff(periods=periods) / data.shift(periods=periods, freq=freq)
```

File: python/cudf/cudf/core/dataframe.py [hunk 2]
```diff
@@ -6329,6 +6329,45 @@ def explode(self, column, ignore_index=False):
...
+            raise NotImplementedError("limit parameter not supported yet.")
+        if freq is not None:
+            raise NotImplementedError("freq parameter not supported yet.")
+        elif fill_method not in {"ffill", "pad", "bfill", "backfill"}:
+            raise ValueError(
+                "fill_method must be one of 'ffill', 'pad', "
+                "'bfill', or 'backfill'."
+            )
+
+        data = self.fillna(method=fill_method, limit=limit)
```
Comment (karthikeyann) [inline python/cudf/cudf/core/dataframe.py:6366]: Does this always create a copy? is it possible to avoid copy by checking something like `has_na`?

Comment (beckernick) [reply python/cudf/cudf/core/dataframe.py:6366]: Could that perhaps happen upstream in the `fillna` function? I could imagine a short circuit based on the null_count

Is there any ecosystem behavior that depends on `fillna` always returning a copy, even if there are no nulls?

File: python/cudf/cudf/core/dataframe.py [hunk 3]
```diff
@@ -6364,9 +6364,10 @@ def pct_change(
             )
 
         data = self.fillna(method=fill_method, limit=limit)
-        data_diff = data.diff(periods=periods)  # need to implem. diff method
-        change = data_diff / data.shift(periods=periods, freq=freq)
-        return change
+
+        return data.diff(periods=periods) / data.shift(
```
Comment (skirui-source) [inline python/cudf/cudf/core/dataframe.py:6368]: dependent on PR (https://github.com/rapidsai/cudf/pull/9817) - work in progress

File: python/cudf/cudf/core/dataframe.py (hunk 2 above)
Comment (skirui-source) [reply python/cudf/cudf/core/dataframe.py:6365]: @karthikeyann @beckernick @shwina According to this thread , looks like there's still pending/ ongoing discussions about optimizing `fillna() `
I was wondering if we can proceed with my current code as is, then optimize in a follow-up PR?

**Review by karthikeyann — APPROVED** [review]

File: python/cudf/cudf/tests/test_dataframe.py [hunk 4]
```diff
@@ -9177,3 +9177,24 @@ def test_dataframe_rename_duplicate_column():
...
+@pytest.mark.parametrize(
+    "data",
+    [
+        np.random.RandomState(seed=10).randint(-50, 50, (25, 30)),
+        np.random.RandomState(seed=10).random_sample((4, 4)),
+        np.array([1.123, 2.343, 5.890, 0.0]),
+        {"a": [1.123, 2.343, np.nan, np.nan], "b": [None, 3, 9.08, None]},
+    ],
+)
+@pytest.mark.parametrize("periods", (-2, -1, 0, 1, 2))
```
Comment (isVoid) [inline python/cudf/cudf/tests/test_dataframe.py:9191]: Let's be consistent in using `list` to specify parameters. And `1` shifts seems too trivial if `2` is tested. Let's also add tests that goes beyond the length of the data (similar to what you did in `diff`).

```suggestion
@pytest.mark.parametrize("periods", [-5, -2, 0, 2, 5])
```

**Review by isVoid — APPROVED** [review]
I don't have much to add here. I think after this we can start thinking about moving `diff` and `pct_change` to `indexed_frame` so that we get both function for `Series` for free.

**Comment by skirui-source** [comment]
@gpucibot merge

**Pull request merged**